interfaces/prompting/requestprompts/requestprompts_test.go

```diff
@@ -1070,3 +1079,162 @@ func (s *requestpromptsSuite) TestPromptMarshalJSON(c *C) {
 
 	c.Assert(string(marshalled), Equals, string(expectedJSON))
 }
+
+func (s *requestpromptsSuite) TestPromptExpiration(c *C) {
+	initialTimeout := 25 * time.Millisecond
+	activityTimeout := 50 * time.Millisecond
+	restore := requestprompts.MockInitialTimeout(initialTimeout)
+	defer restore()
+	restore = requestprompts.MockActivityTimeout(activityTimeout)
+	defer restore()
+
+	replyChan := make(chan notify.FilePermission, 1)
+	restore = requestprompts.MockSendReply(func(listenerReq *listener.Request, allowedPermission any) error {
+		allowedFilePermission, ok := allowedPermission.(notify.FilePermission)
+		c.Assert(ok, Equals, true)
+		replyChan <- allowedFilePermission
+		return nil
+	})
+	defer restore()
+
+	metadata := &prompting.Metadata{
+		User:      s.defaultUser,
+		Snap:      "firefox",
+		Interface: "home",
+	}
+	path := "/home/test/foo"
+	requestedPermissions := []string{"read", "write", "execute"}
+	remainingPermissions := []string{"write", "execute"}
+
+	// Should only have one at a time, but leave space for 2 in case expiration
+	// happens early before creation notice has been received.
+	noticeChan := make(chan noticeInfo, 2)
+	pdb, err := requestprompts.New(func(userID uint32, promptID prompting.IDType, data map[string]string) error {
+		c.Assert(userID, Equals, s.defaultUser)
+		noticeChan <- noticeInfo{
+			promptID: promptID,
+			data:     data,
+		}
+		return nil
+	})
+	c.Assert(err, IsNil)
+	defer pdb.Close()
+
+	// Add prompt
+	listenerReq := &listener.Request{}
+	prompt, merged, err := pdb.AddOrMerge(metadata, path, requestedPermissions, remainingPermissions, listenerReq)
+	c.Assert(err, IsNil)
+	c.Assert(merged, Equals, false)
+	checkCurrentNotices(c, noticeChan, prompt.ID, nil)
+
+	// Check that prompt has not immediately expired
+	checkNoNotices(c, noticeChan)
+	checkNoReply(c, replyChan)
+
+	// Prompt should expire after initialTimeout
+	time.Sleep(initialTimeout)
+	checkCurrentNotices(c, noticeChan, prompt.ID, map[string]string{"resolved": "expired"})
```

Comment:
It would be less fragile to not use time for this. Instead of mocking the milliseconds, mock the after func. Mock the entire "timer" so that you can freely trigger it when you need to and not rely on actual time in any of the tests.